File: src/utils/postStore.js
// ── Post store: queries and mutations ─────────────────────────
// All database operations for blog posts.

const { buildExcerpt } = require('./excerpt');
const { renderMarkdown } = require('./markdown');
const {
    STATUSES, getCollection, normaliseTags, computeSlug,
    formatDateValue, ensureSlugAvailable, mapListDocument,
} = require('./postHelpers');

async function listPosts({ includeDrafts = false, search, tag } = {}) {
    const query = {};
    if (!includeDrafts) query.status = STATUSES.PUBLISHED;
    if (tag) query.tags = tag.toLowerCase();

    const docs = await getCollection()
        .find(query, { projection: { markdown: 0 }, sort: { publishedAt: -1, updatedAt: -1 } })
        .toArray();

    let posts = docs.map(mapListDocument);
    if (search) {
        const low = search.toLowerCase();
        posts = posts.filter((p) =>
            p.title.toLowerCase().includes(low) || (p.excerpt || '').toLowerCase().includes(low));
    }
    return posts;
}

async function getPostBySlug(slug, { includeDrafts = false } = {}) {
    const query = { slug: slug.trim().toLowerCase() };
    if (!includeDrafts) query.status = STATUSES.PUBLISHED;
    const doc = await getCollection().findOne(query);
    if (!doc) return null;
    return {
        slug: doc.slug, title: doc.title, status: doc.status,
        tags: doc.tags || [],
        createdAt: formatDateValue(doc.createdAt),
        updatedAt: formatDateValue(doc.updatedAt),
        publishedAt: formatDateValue(doc.publishedAt),
        excerpt: doc.excerpt || '',
        markdown: doc.markdown || '',
        html: renderMarkdown(doc.markdown || ''),
        heroImage: doc.heroImage || null,
        attachments: doc.attachments || [],
    };
}

async function createPost({ title, markdown, tags, status = STATUSES.DRAFT, slug: desiredSlug }) {
    const slug = computeSlug(title, desiredSlug);
    await ensureSlugAvailable(slug);
    const normalizedStatus = status === STATUSES.PUBLISHED ? STATUSES.PUBLISHED : STATUSES.DRAFT;
    const now = new Date();
    await getCollection().insertOne({
        slug, title, markdown: markdown || '',
        tags: normaliseTags(tags), status: normalizedStatus,
        createdAt: now, updatedAt: now,
        publishedAt: normalizedStatus === STATUSES.PUBLISHED ? now : null,
        excerpt: buildExcerpt(markdown),
        heroImage: null, attachments: [],
    });
    return getPostBySlug(slug, { includeDrafts: true });
}

async function updatePost(originalSlug, { title, markdown, tags, status, slug: desiredSlug }) {
    const col = getCollection();
    const existing = await col.findOne({ slug: originalSlug });
    if (!existing) { const e = new Error(`Post "${originalSlug}" not found`); e.status = 404; throw e; }

    const nextStatus = status === STATUSES.PUBLISHED ? STATUSES.PUBLISHED
        : status === STATUSES.DRAFT ? STATUSES.DRAFT : existing.status;
    const newTitle = title || existing.title;
    const slug = computeSlug(newTitle, desiredSlug || existing.slug);
    const md = typeof markdown === 'string' ? markdown : existing.markdown || '';
    const now = new Date();
    if (slug !== existing.slug) await ensureSlugAvailable(slug, existing._id);

    await col.updateOne({ _id: existing._id }, { $set: {
        slug, title: newTitle, markdown: md,
        tags: typeof tags !== 'undefined' ? normaliseTags(tags) : existing.tags || [],
        status: nextStatus, updatedAt: now,
        publishedAt: nextStatus === STATUSES.PUBLISHED ? (existing.publishedAt || now) : null,
        excerpt: buildExcerpt(md),
    }});
    return getPostBySlug(slug, { includeDrafts: true });
}

async function deletePost(slug) {
    const result = await getCollection().deleteOne({ slug });
    if (result.deletedCount === 0) {
        const e = new Error(`Post "${slug}" not found`); e.status = 404; throw e;
    }
}

async function collectTags() {
    const pipeline = [
        { $unwind: '$tags' },
        { $group: { _id: '$tags', count: { $sum: 1 } } },
        { $sort: { count: -1, _id: 1 } },
    ];
    return (await getCollection().aggregate(pipeline).toArray())
        .map((e) => ({ tag: e._id, count: e.count }));
}

module.exports = { STATUSES, listPosts, getPostBySlug, createPost, updatePost, deletePost, collectTags };

File: post_store.py
from datetime import datetime, timezone
from slugify import slugify
from db import get_db
from excerpt import build_excerpt
from markdown_render import render_markdown

PUBLISHED, DRAFT = "published", "draft"
_COL = "posts"


def _col():
    return get_db()[_COL]


def _tags(raw) -> list[str]:
    if not raw:
        return []
    parts = raw if isinstance(raw, list) else str(raw).split(",")
    return [t.strip().lower() for t in parts if t.strip()]


def _slug(title: str, desired: str | None) -> str:
    return slugify(desired or title, separator="-") or slugify(title, separator="-")


def _fmt(v) -> str | None:
    if not v:
        return None
    return v.isoformat() if isinstance(v, datetime) else str(v)


def _map(doc: dict) -> dict:
    return {k: doc.get(k) for k in ("slug", "title", "status", "tags", "excerpt", "heroImage")} | {
        "createdAt": _fmt(doc.get("createdAt")),
        "updatedAt": _fmt(doc.get("updatedAt")),
        "publishedAt": _fmt(doc.get("publishedAt")),
    }


async def list_posts(include_drafts=False, tag=None, search=None) -> list[dict]:
    q = {} if include_drafts else {"status": PUBLISHED}
    if tag:
        q["tags"] = tag.lower()
    docs = await _col().find(q, {"markdown": 0}).sort([("publishedAt", -1), ("updatedAt", -1)]).to_list(None)
    posts = [_map(d) for d in docs]
    if search:
        low = search.lower()
        posts = [p for p in posts if low in p["title"].lower() or low in (p.get("excerpt") or "").lower()]
    return posts


async def get_post(slug: str, include_drafts=False) -> dict | None:
    q = {"slug": slug.strip().lower()}
    if not include_drafts:
        q["status"] = PUBLISHED
    doc = await _col().find_one(q)
    if not doc:
        return None
    return _map(doc) | {"markdown": doc.get("markdown", ""), "html": render_markdown(doc.get("markdown", ""))}


async def create_post(title: str, markdown: str, tags, status: str, slug: str | None = None) -> dict:
    s = _slug(title, slug)
    if await _col().find_one({"slug": s}, {"_id": 1}):
        raise ValueError(f'Slug "{s}" already exists')
    now = datetime.now(timezone.utc)
    status = PUBLISHED if status == PUBLISHED else DRAFT
    await _col().insert_one({
        "slug": s, "title": title, "markdown": markdown or "",
        "tags": _tags(tags), "status": status, "excerpt": build_excerpt(markdown),
        "heroImage": None, "attachments": [],
        "createdAt": now, "updatedAt": now,
        "publishedAt": now if status == PUBLISHED else None,
    })
    return await get_post(s, include_drafts=True)  # type: ignore[return-value]


async def update_post(orig_slug: str, *, title=None, markdown=None, tags=None, status=None, slug=None) -> dict:
    doc = await _col().find_one({"slug": orig_slug})
    if not doc:
        raise LookupError(f'Post "{orig_slug}" not found')
    new_title = title or doc["title"]
    new_slug = _slug(new_title, slug or doc["slug"])
    new_status = PUBLISHED if status == PUBLISHED else (DRAFT if status == DRAFT else doc["status"])
    new_md = markdown if markdown is not None else doc.get("markdown", "")
    now = datetime.now(timezone.utc)
    if new_slug != orig_slug and await _col().find_one({"slug": new_slug, "_id": {"$ne": doc["_id"]}}, {"_id": 1}):
        raise ValueError(f'Slug "{new_slug}" already exists')
    await _col().update_one({"_id": doc["_id"]}, {"$set": {
        "slug": new_slug, "title": new_title, "markdown": new_md,
        "tags": _tags(tags) if tags is not None else doc.get("tags", []),
        "status": new_status, "updatedAt": now,
        "publishedAt": (doc.get("publishedAt") or now) if new_status == PUBLISHED else None,
        "excerpt": build_excerpt(new_md),
    }})
    return await get_post(new_slug, include_drafts=True)  # type: ignore[return-value]


async def delete_post(slug: str) -> None:
    r = await _col().delete_one({"slug": slug})
    if r.deleted_count == 0:
        raise LookupError(f'Post "{slug}" not found')
